group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeAssignor.java

```diff
@@ -65,191 +94,233 @@ public String name() {
+     * Assignment will be co-partitioned when all the topics have an equal number of partitions.
      */
-    private Map<Uuid, Collection<String>> membersPerTopic(
-        final GroupSpec groupSpec,
-        final SubscribedTopicDescriber subscribedTopicDescriber
-    ) {
-        Map<Uuid, Collection<String>> membersPerTopic = new HashMap<>();
-
-        if (groupSpec.subscriptionType().equals(HOMOGENEOUS)) {
-            Collection<String> allMembers = groupSpec.memberIds();
-            Collection<Uuid> topics = groupSpec.memberSubscription(groupSpec.memberIds().iterator().next())
-                .subscribedTopicIds();
-
-            for (Uuid topicId : topics) {
-                if (subscribedTopicDescriber.numPartitions(topicId) == -1) {
-                    throw new PartitionAssignorException("Member is subscribed to a non-existent topic");
-                }
-                membersPerTopic.put(topicId, allMembers);
+    private GroupAssignment assignHomogeneousGroup(
+        GroupSpec groupSpec,
+        SubscribedTopicDescriber subscribedTopicDescriber
+    ) throws PartitionAssignorException {
+        List<String> memberIds = sortMemberIds(groupSpec);
+
+        MemberSubscription subs = groupSpec.memberSubscription(memberIds.get(0));
+        Set<Uuid> subscribedTopics = new HashSet<>(subs.subscribedTopicIds());
+        List<TopicMetadata> topics = new ArrayList<>(subscribedTopics.size());
+        int numMembers = groupSpec.memberIds().size();
+
+        for (Uuid topicId : subscribedTopics) {
+            int numPartitions = subscribedTopicDescriber.numPartitions(topicId);
+            if (numPartitions == -1) {
+                throw new PartitionAssignorException("Member is subscribed to a non-existent topic");
             }
-        } else {
-            groupSpec.memberIds().forEach(memberId -> {
-                Collection<Uuid> topics = groupSpec.memberSubscription(memberId).subscribedTopicIds();
-                for (Uuid topicId : topics) {
-                    if (subscribedTopicDescriber.numPartitions(topicId) == -1) {
-                        throw new PartitionAssignorException("Member is subscribed to a non-existent topic");
-                    }
-                    membersPerTopic
-                        .computeIfAbsent(topicId, k -> new ArrayList<>())
-                        .add(memberId);
-                }
-            });
+            TopicMetadata m = TopicMetadata.create(
+                topicId,
+                numPartitions,
+                numMembers
+            );
+            topics.add(m);
         }
 
-        return membersPerTopic;
+        Map<String, MemberAssignment> assignments = new HashMap<>((int) ((groupSpec.memberIds().size() / 0.75f) + 1));
+
+        for (String memberId : memberIds) {
+            Map<Uuid, Set<Integer>> assignment = new HashMap<>((int) ((subscribedTopics.size() / 0.75f) + 1));
```

Comment:
nit: Would it make sense to define a variable for `(int) ((subscribedTopics.size() / 0.75f) + 1)` as we reuse it in every iteration?